engine/src/main/java/org/hibernate/validator/internal/engine/cascading/ValueExtractors.java

```diff
@@ -64,13 +65,15 @@ public ValueExtractors(Iterable<ValueExtractor<?>> externalExtractors) {
 			for ( TypeVariable<?> cascadingValueTypeParameter : cascadingTypeParametersOfValueType ) {
 				TypeVariable<?> cascadingParameterBoundToExtractorType = bind( cascadingValueTypeParameter, bindingsForExtractorType );
 				if ( cascadingParameterBoundToExtractorType.equals( extractorDescriptor.extractedTypeParameter() ) ) {
-					cascadedValueExtractors.add( extractorDescriptor.getValueExtractor() );
+					// TODO implement selection of most specific extractor per requested type parameter
+					if ( !cascadedValueExtractors.containsKey( cascadingValueTypeParameter ) ) {
+						cascadedValueExtractors.put( cascadingValueTypeParameter, extractorDescriptor.getValueExtractor() );
+					}
 				}
 			}
 		}
 
-		// TODO implement selection of most specific extractor per requested type parameter
-		return cascadedValueExtractors.size() > 1 ? Collections.singletonList( cascadedValueExtractors.iterator().next() ) : cascadedValueExtractors;
+		return new ArrayList<>( cascadedValueExtractors.values() );
```

Comment:
It currently is important, but eventually it shouldn't be. In the end I think we need to keep the most specific extractor per type parameter, e.g. if we have an extractor for `Iterable<@ExtractedValue E` and one for `List<@ExtractedValue E` and the validated value is an `ArrayList`, only keep the latter. But I didn't get there yet. Currently it will simply keep the first extractor found per type variable, hence the order of insertion in the constructor is relevant. It's a hack to make things work a bit.